### PR TITLE
PERF: performance regression in Series.asof

### LINKED ISSUE #14461: pd.Series.asof performance regression: 14x slower
#### pd.Series.asof takes 14x longer in 0.19.0 than 0.18.1

```python
In[14]: pd.__version__
Out[14]: '0.18.1'
In[15]: s = pd.Series([1, 1, 1, np.nan, np.nan])
In[16]: %timeit s.asof(4)
The slowest run took 8.25 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 8.96 µs per loop

In[8]: pd.__version__
Out[8]: '0.19.0'
In[9]: s = pd.Series([1, 1, 1, np.nan, np.nan])
In[10]: %timeit s.asof(4)
10000 loops, best of 3: 125 µs per loop
```

### PR DESCRIPTION
**Pull request by laudney**
Fix performance regression in Series.asof by avoiding pre-computing nulls and returning value by indexing the underlying ndarray.
- [x] closes #14461 
- [x] add asv benchmark

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
needs an asv on several different cases
this will be very slow with lots of nulls

**Comment by laudney** [comment]
@jreback you mean we should always pre-compute nulls for DataFrame?

**Comment by jreback** [comment]
the answer is probably yes
but you need some perf testing in several scenarios to see that is an asv

**Comment by laudney** [comment]
@jreback what's asv?

**Comment by jreback** [comment]
http://pandas.pydata.org/pandas-docs/stable/contributing.html#running-the-performance-test-suite

**Comment by jorisvandenbossche** [comment]
There are some benchmarks for the series case already: https://github.com/pandas-dev/pandas/blob/master/asv_bench/benchmarks/timeseries.py#L283. We should indeed add for DataFrame.

For Series I suspect that the approach in this PR will also be faster with lots of NaNs, as the indexing of Series is much slower that the isnull checking on single values. For DataFrame not directly sure, so there is may make sense to precompute the NaNs (should be checked).

**Comment by laudney** [comment]
asv results as the pull request stands now:

```python
    before     after       ratio
  [233d51dd] [4836cdfd]
+  232.75μs   263.67μs      1.13  timeseries.timeseries_custom_bmonthend_incr.time_timeseries_custom_bmonthend_incr
+   14.06ms    15.68ms      1.11  timeseries.datetime_algorithm.time_dti_factorize
+   24.23μs    26.87μs      1.11  timeseries.timeseries_year_apply.time_timeseries_year_apply
+  338.21μs   374.77μs      1.11  timeseries.timeseries_custom_bmonthend_decr_n.time_timeseries_custom_bmonthend_decr_n
+  687.57ms   761.01ms      1.11  timeseries.timeseries_semi_month_offset.time_semi_month_begin_decr_rng
-   27.32ms    24.70ms      0.90  timeseries.timeseries_iter_periodindex_preexit.time_timeseries_iter_periodindex_preexit
-    3.12ms     2.77ms      0.89  timeseries.dataframe_resample_mean_numpy.time_dataframe_resample_mean_numpy
-    3.45ms     2.99ms      0.87  timeseries.dataframe_resample_max_string.time_dataframe_resample_max_string
-  314.65μs   271.87μs      0.86  timeseries.datetimeindex_add_offset.time_datetimeindex_add_offset
-  409.15μs    24.12μs      0.06  timeseries.timeseries_asof_single.time_timeseries_asof_single
```

**Comment by jorisvandenbossche** [comment]
@laudney You can selectively run only the asof ones with `-b asof` (in this case that is enough, as you do not touch code related to other functionality).

**Comment by laudney** [comment]
@jorisvandenbossche I did -b ^timeseries

**Comment by laudney** [comment]
asv ... -b asof

```
    before     after       ratio
  [233d51dd] [4836cdfd]
-  454.34μs    22.44μs      0.05  timeseries.timeseries_asof_single.time_timeseries_asof_single
```

**Comment by jreback** [comment]
need several more asvs

**Comment by laudney** [comment]
I've just updated the PR to include 3 new asv tests for DataFrame.asof. Please note that it shows it's almost always better *not* to pre-compute nulls even for DataFrame. In the worst case of large number of consecutive NAs, searching backward one row at a time is still as fast as the pre-computation.

```
· Running 18 total benchmarks (2 commits * 1 environments * 9 benchmarks)
[  0.00%] · For pandas commit hash 4836cdfd:
[  0.00%] ·· Building for virtualenv-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-scipy-sqlalchemy-tables-xlrd-xlsxwriter-xlwt...
[  0.00%] ·· Benchmarking virtualenv-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-scipy-sqlalchemy-tables-xlrd-xlsxwriter-xlwt
[  5.56%] ··· Running join_merge.merge_asof_by_int.time_merge_asof_by_int                                                           30.01ms
[ 11.11%] ··· Running join_merge.merge_asof_by_object.time_merge_asof_by_object                                                     50.94ms
[ 16.67%] ··· Running join_merge.merge_asof_noby.time_merge_asof_noby                                                               13.36ms
[ 22.22%] ··· Running timeseries.timeseries_asof.time_timeseries_asof                                                                9.95ms
[ 27.78%] ··· Running timeseries.timeseries_asof_nan.time_timeseries_asof_nan                                                        9.96ms
[ 33.33%] ··· Running timeseries.timeseries_asof_single.time_timeseries_asof_single                                                 22.72μs
[ 38.89%] ··· Running timeseries.timeseries_dataframe_asof.time_timeseries_asof                                                    598.80ms
[ 44.44%] ··· Running timeseries.timeseries_dataframe_asof_nan.time_timeseries_asof_nan                                            578.84ms
[ 50.00%] ··· Running timeseries.timeseries_dataframe_asof_single.time_timeseries_asof_single                                      143.14μs
[ 50.00%] · For pandas commit hash 233d51dd:
[ 50.00%] ·· Building for virtualenv-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-scipy-sqlalchemy-tables-xlrd-xlsxwriter-xlwt...
[ 50.00%] ·· Benchmarking virtualenv-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-scipy-sqlalchemy-tables-xlrd-xlsxwriter-xlwt
[ 55.56%] ··· Running join_merge.merge_asof_by_int.time_merge_asof_by_int                                                           29.21ms
[ 61.11%] ··· Running join_merge.merge_asof_by_object.time_merge_asof_by_object                                                     48.55ms
[ 66.67%] ··· Running join_merge.merge_asof_noby.time_merge_asof_noby                                                               14.08ms
[ 72.22%] ··· Running timeseries.timeseries_asof.time_timeseries_asof                                                               10.09ms
[ 77.78%] ··· Running timeseries.timeseries_asof_nan.time_timeseries_asof_nan                                                        9.85ms
[ 83.33%] ··· Running timeseries.timeseries_asof_single.time_timeseries_asof_single                                                454.21μs
[ 88.89%] ··· Running timeseries.timeseries_dataframe_asof.time_timeseries_asof                                                    557.98ms
[ 94.44%] ··· Running timeseries.timeseries_dataframe_asof_nan.time_timeseries_asof_nan                                            572.75ms
[100.00%] ··· Running timeseries.timeseries_dataframe_asof_single.time_timeseries_asof_single                                      177.38ms
  before     after       ratio
  [233d51dd] [4836cdfd]
-  454.21μs    22.72μs      0.05  timeseries.timeseries_asof_single.time_timeseries_asof_single
-  177.38ms   143.14μs      0.00  timeseries.timeseries_dataframe_asof_single.time_timeseries_asof_single

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

File: asv_bench/benchmarks/timeseries.py [hunk 1]
```diff
@@ -335,6 +335,48 @@ def time_timeseries_asof_single(self):
         self.ts.asof(self.dates[0])
 
 
+class timeseries_dataframe_asof(object):
+    goal_time = 0.2
+
+    def setup(self):
+        self.N = 10000
```
Comment (jreback) [inline asv_bench/benchmarks/timeseries.py]: these are too small make 100000

put nans at beginning and another one at the end

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by laudney** [comment]
I've updated my pull request code to refactor existing asv tests for Series.asof and add new ones for DataFrame.asof.

The main findings are:

1. Avoid pre-computation of nulls for Series if *where* is a scalar because the *while* loop is always faster than the pre-computation.
2. Always pre-compute nulls for DataFrames. The *while* loop is extremely expensive. Given pre-computation is always available, even if *where* is a scalar, should avoid the *while* loop and use the same code path as when *where* is list-like.

asv results below:

```
    before     after       ratio
  [233d51dd] [4cd64641]
+   15.63ms    20.19ms      1.29  timeseries.timeseries_dataframe_asof.time_asof_single
-  429.14μs    22.11μs      0.05  timeseries.timeseries_asof.time_asof_single
-     1.28s    21.10ms      0.02  timeseries.timeseries_dataframe_asof.time_asof_nan_single
-     1.24s     3.56ms      0.00  timeseries.timeseries_asof.time_asof_nan_single
```

**Review by jreback — CHANGES_REQUESTED** [review]
can u add your explanation of perf in a comment
with a ref to this pr number in a comment as well 

lgtm otherwise

**Comment by laudney** [comment]
@jreback I've added comment in the PR. Please check and let me know. Off to bed.

**Comment by laudney** [comment]
I will add two more asv tests and more detailed comment tonight. I will ping you when ready.

**Comment by jreback** [comment]
lgtm. pls add a whatsnew entry in the performance section. ping on green.

**Comment by laudney** [comment]
@jreback @jorisvandenbossche I've updated the PR: more comment and more asv tests. It's ready. Please review and merge.

impact as below:

```
-   87.33μs    18.44μs      0.21  timeseries.timeseries_asof.time_asof_single_early
-  390.14μs    21.42μs      0.05  timeseries.timeseries_asof.time_asof_single
-     1.20s     3.34ms      0.00  timeseries.timeseries_asof.time_asof_nan_single
-     1.26s    21.55ms      0.02  timeseries.timeseries_dataframe_asof.time_asof_nan_single
-   14.65ms    20.11μs      0.00  timeseries.timeseries_dataframe_asof.time_asof_single_early
```

**Comment by jreback** [comment]
@laudney pls add a whatsnew note

**Comment by laudney** [comment]
@jreback whatsnew updated for 0.19.1

**Comment by laudney** [comment]
@jreback @jorisvandenbossche It's all green on my side. Let me know if you need anything else.

**Comment by jorisvandenbossche** [comment]
@laudney Thanks!

**Pull request merged**